tests/check_framework/test_model_checks.py

```diff
@@ -50,6 +50,31 @@ class Meta:
             )
         ])
 
+    @modify_settings(INSTALLED_APPS={'append': 'basic'})
+    @isolate_apps('basic', 'check_framework', kwarg_name='apps')
+    def test_collision_across_apps_special(self, apps):
+        constraint = models.CheckConstraint(check=models.Q(id__gt=0), name='%(app_label)s_%(class)s_foo')
+        index = models.Index(fields=['id'], name='%(app_label)s_%(class)s_index')
+
+        class Model1(models.Model):
+            class Meta:
+                app_label = 'basic'
+                constraints = [constraint]
+                indexes = [index]
+
+        class Model2(models.Model):
+            class Meta:
+                app_label = 'check_framework'
+                constraints = [constraint]
+                indexes = [index]
+
+        self.assertEqual(checks.run_checks(app_configs=apps.get_app_configs()), [])
```

Comment:
In `tests/check_framework` we should test only system checks. Test for names can be added to `constraints/tests.py` and `indexes/tests.py`.